iceoryx_dust/vocabulary/include/iox/span.hpp

```diff
@@ -0,0 +1,429 @@
+using enable_if_conversion_allowed_t =
+    std::enable_if_t<(ToExtent == DYNAMIC_EXTENT || ToExtent == FromExtent) && is_convertible_t<From, To>::value>;
+
+// SFINAE Verify it the passed array can be converted to a span<T>.
+template <typename Array, typename T, uint64_t Extent>
+using enable_if_compatible_array_t =
+    std::enable_if_t<(Extent == DYNAMIC_EXTENT || Extent == detail::extent_t<Array>::value)
+                     && container_has_convertible_data_t<Array, T>::value>;
+
+// SFINAE Verify it the passed container can be converted to a span<T>.
+template <typename Container, typename T>
+using is_compatible_container_t = conjunction<is_not_span_t<Container>,
+                                              iox::is_not_iox_array_t<Container>,
+                                              iox::is_not_c_array_t<Container>,
+                                              container_has_convertible_data_t<Container, T>,
+                                              is_integral_sized_container_t<Container>>;
+
+template <typename Container, typename T>
+using enable_if_compatible_container_t = std::enable_if_t<is_compatible_container_t<Container, T>::value>;
+
+template <typename Container, typename T, uint64_t Extent>
+using enable_if_compatible_dynamic_container_t =
+    std::enable_if_t<is_compatible_container_t<Container, T>::value && Extent == DYNAMIC_EXTENT>;
+
+template <uint64_t Extent>
+class span_storage
+{
+  public:
+    constexpr explicit span_storage(uint64_t) noexcept
+    {
+    }
+    constexpr uint64_t size() const noexcept
+    {
+        return Extent;
+    }
+};
+
+template <>
+struct span_storage<DYNAMIC_EXTENT>
+{
+    constexpr explicit span_storage(uint64_t size) noexcept
+        : m_size(size)
+    {
+    }
+    constexpr uint64_t size() const noexcept
+    {
+        return m_size;
+    }
+
+    span_storage(const span_storage& other) noexcept = default;
+    span_storage& operator=(const span_storage& other) noexcept = default;
+
+    span_storage(span_storage&& other) noexcept
+    {
+        *this = std::move(other);
+    }
+    span_storage& operator=(span_storage&& other) noexcept
+    {
+        if (this != &other)
+        {
+            m_size = other.m_size;
+            other.m_size = 0;
+        }
+        return *this;
+    }
+    ~span_storage() noexcept = default;
+
+  private:
+    uint64_t m_size{0};
+};
+
+template <typename T>
+constexpr T* to_address(T* p) noexcept
+{
+    static_assert(!std::is_function<T>::value, "Error: T must not be a function type.");
+    return p;
+}
+
+template <typename Ptr>
+constexpr auto to_address(const Ptr& p) noexcept -> decltype(std::pointer_traits<Ptr>::to_address(p))
+{
+    return std::pointer_traits<Ptr>::to_address(p);
+}
+
+template <typename Ptr, typename... None>
+constexpr auto to_address(const Ptr& p, None...) noexcept
+{
+    return to_address(p.operator->());
+}
+
+} // namespace detail
```

Comment:
Quite a lot of the functions in `detail` have generic names which could lead to issues later on. I'm myself not sure what would be the best to do. Prefix, another namespace or just keep it as is?

Reply:
I think for now leave it as it is, but I would move the generic things outside in future. Is that ok from your side?